core/src/main/java/org/apache/iceberg/io/ContentCache.java

```diff
@@ -0,0 +1,300 @@
+    this.expireAfterAccessMs = expireAfterAccessMs;
+    this.maxTotalBytes = maxTotalBytes;
+    this.maxContentLength = maxContentLength;
+
+    Caffeine<Object, Object> builder = Caffeine.newBuilder();
+    if (expireAfterAccessMs > 0) {
+      builder = builder.expireAfterAccess(Duration.ofMillis(expireAfterAccessMs));
+    }
+
+    this.cache =
+        builder
+            .maximumWeight(maxTotalBytes)
+            .weigher(
+                (Weigher<String, CacheEntry>)
+                    (key, value) -> (int) Math.min(value.length, Integer.MAX_VALUE))
+            .recordStats()
+            .build();
+  }
+
+  public long expireAfterAccess() {
+    return expireAfterAccessMs;
+  }
+
+  public long maxContentLength() {
+    return maxContentLength;
+  }
+
+  public long maxTotalBytes() {
+    return maxTotalBytes;
+  }
+
+  public CacheStats stats() {
+    return cache.stats();
+  }
+
+  public CacheEntry get(String key, Function<String, CacheEntry> mappingFunction) {
+    return cache.get(key, mappingFunction);
+  }
+
+  public CacheEntry getIfPresent(String location) {
+    return cache.getIfPresent(location);
+  }
+
+  public InputFile tryCache(FileIO io, String location, long length) {
+    if (length <= maxContentLength) {
+      return new CachingInputFile(this, io, location, length);
+    }
+    return io.newInputFile(location, length);
+  }
+
+  public void invalidate(String key) {
+    cache.invalidate(key);
+  }
+
+  public void invalidateAll() {
+    cache.invalidateAll();
+  }
+
+  public void cleanUp() {
+    cache.cleanUp();
+  }
+
+  public long estimatedCacheSize() {
+    return cache.estimatedSize();
+  }
+
+  @Override
+  public String toString() {
+    return getClass().getSimpleName()
+        + '{'
+        + "expireAfterAccessMs="
+        + expireAfterAccessMs
+        + ", "
+        + "maxContentLength="
+        + maxContentLength
+        + ", "
+        + "maxTotalBytes="
+        + maxTotalBytes
+        + ", "
+        + "cacheStats="
+        + cache.stats()
+        + '}';
+  }
+
+  private static class CacheEntry {
+    private final long length;
+    private final List<ByteBuffer> buffers;
+
+    private CacheEntry(long length, List<ByteBuffer> buffers) {
+      this.length = length;
+      this.buffers = buffers;
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final FileIO io;
+    private final String location;
+    private final long length;
+    private InputFile fallbackInputFile = null;
+
+    private CachingInputFile(ContentCache cache, FileIO io, String location, long length) {
+      this.contentCache = cache;
+      this.io = io;
+      this.location = location;
+      this.length = length;
+    }
+
+    private InputFile wrappedInputFile() {
+      if (fallbackInputFile == null) {
+        fallbackInputFile = io.newInputFile(location, length);
+      }
+      return fallbackInputFile;
```

Comment:
How long do we keep this reference for the `InputFile`. Does this have any reference for important resources?

Reply:
This `InputFile` reference is kept as long as the enclosing `CachingInputFile`. The `CachingInputFile` itself is not cached. Only the byte buffers of the input stream are cached on the first call to `CachingInputFile#cachedStream()`

Reply:
Sounds reasonable to me